iroh-net/src/derp.rs

```diff
@@ -28,3 +36,109 @@ pub use self::server::{
     ClientConnHandler, MaybeTlsStream as MaybeTlsStreamServer, PacketForwarderHandler, Server,
 };
 pub use self::types::{MeshKey, PacketForwarder};
+
+/// A URL identifying a DERP server.
+///
+/// This is but a wrapper around [`Url`], with a few custom tweaks:
+///
+/// - A DERP URL is never a relative URL, so an implicit `.` is added at the end of the
+///   domain name if missing.
+///
+/// - [`fmt::Debug`] is implemented so it prints the URL rather than the URL struct fields.
+///   Useful when logging e.g. `Option<DerpUrl>`.
+///
+/// To create a [`DerpUrl`] use the `From<Url>` implementation.
+#[derive(
+    Clone, derive_more::Display, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize,
+)]
+pub struct DerpUrl(Url);
+
+impl From<Url> for DerpUrl {
+    fn from(mut url: Url) -> Self {
+        if let Some(domain) = url.domain() {
+            if !domain.ends_with('.') {
+                let domain = String::from(domain) + ".";
+
+                // This can fail, though it is unlikely the resulting URL is usable as a
+                // DERP URL, probably it has the wrong scheme or is not a base URL or the
+                // like.  We don't do full URL validation however, so just silently leave
+                // this bad URL in place.  Something will fail later.
+                url.set_host(Some(&domain)).ok();
+            }
+        }
+        Self(url)
+    }
+}
+
+impl FromStr for DerpUrl {
+    // Be aware, we are re-exporting another crate's public type in our API.
+    type Err = url::ParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let inner = Url::from_str(s)?;
+        Ok(DerpUrl(inner))
+    }
+}
+
+/// Dereference to the wrapped [`Url`].
+///
+/// Note that [`DerefMut`] is not implemented on purpose, so this type has more flexibility
+/// to change the inner later.
+///
+/// [`DerefMut`]: std::ops::DerefMut
+impl Deref for DerpUrl {
+    type Target = Url;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl fmt::Debug for DerpUrl {
```

Comment:
missing `Display` impl for `DerpUrl`

Reply:
There is a derive `derive_more::Display`.  I've added a test for that functionality now as well.

Reply:
lol, I was just blind I guess